design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+				},
+			},
+		}
+	}
+	return admission.Allowed("")
+}
+```
+
+### User Experience
+
+**Defining a Usage Resource as part of a Composition:**
+
+We expect a typical usage relationship to be defined as part of the same
+composition as the using and used resources. In this case, the user will
+define the `Usage` resource as part of the composition template.
+
+Let's take [upbound/platform-ref-aws] as an example. In this composition, we
+have a [composition for `XCluster`] which composes `XEKS` composing an EKS cluster,
+and `XServices` composing a set of helm releases running on the cluster. In this
+case, we will add the following resource to the same composition to define a
+usage relationship between the `XServices` and the `XEKS` Composites:
+
+```yaml
+    - name: xservices-uses-xeks
+      base:
+        apiVersion: apiextensions.crossplane.io/v1alpha1
+        kind: Usage
+        spec:
+          of:
+            - apiVersion: aws.platformref.upbound.io/v1alpha1
+              kind: XEKS
+              selector:
+                matchControllerRef: true
+          by:
+            apiVersion: aws.platformref.upbound.io/v1alpha1
+            kind: XServices
+            selector:
+              matchControllerRef: true
+```
+
+**Deleting a Claim or Composite which contains a Usage Relationship:**
+
+**With the (default) background propagation policy**, we expect no different in user
+experience. Users will be able to delete the Claim or Composite without any error. 
+The Claim or Composite will be deleted immediately, and the Kubernetes garbage
+collector will delete the Composed Resources together with the `Usage` resources
+in the background. It will take longer until all Composed Resources are garbage
+collected, however this will not be visible to the user.
+
+**With the foreground propagation policy**, the only difference we expect is
+that the deletion is taking longer than before. The user will be able to delete
+the Claim or Composite without any error. The deletion of the Claim or Composite
+will be blocked until all Composed Resources are deleted as before. It will take
+longer because deletion of the used resource will be blocked until the used
+resource hence `Usage` is deleted.
+
+> One caveat with the garbage collector is the substantial delay it can cause
+during the deletion of composites. This delay caused by the garbage collector's
+use of an exponential backoff when the DELETE API call is rejected. The maximum
+backoff time is 1000 seconds, which means that after no usage of an object is
+left, it could take up to 1000 seconds before the garbage collector attempts to
+delete the resource that was previously blocked. To mitigate this issue, we
+could introduce a mechanism in the controller to re-trigger the deletion of the
+used resource as soon as the Usage resource is deleted.
```

Comment:
The downside of this is that we have to reimplement basically the GC logic of "when not to delete an object". I.e. there is risk that Kube introduces something that stops the GC from deleting an object (think of another ownership mechanism, or even something like this `Usage` idea here in the proposal). Our code won't be aware and too quickly delete an object bypassing GC.

My gut feeling is that this risk is limited as these things don't show up out of nowhere, but api-machinery is slow moving and if it ever happens, we have to revisit our code.

Reply:
Wouldn't the DELETE request that we make be subject to the same processing/restrictions as any other DELETE request? I don't see that we are bypassing GC, we are just asking it to try again, and it may fail for other reasons, which will be outside the scope of our scenario, and we will never try again.

Reply:
My fear is that we are deleting more eagerly than Kube's GC because we miss something in our logic.

Reply:
I believe we are aligned on this in [the other thread](https://github.com/crossplane/crossplane/pull/4215#discussion_r1239579950).